pootle/apps/pootle_fs/files.py

```diff
@@ -0,0 +1,223 @@
+            % (self.path,
+               self.pootle_path,
+               self.store_fs.last_sync_hash,
+               self.store_fs.last_sync_revision))
+
+    def __eq__(self, other):
+        return hash(other) == hash(self)
+
+    @property
+    def exists(self):
+        return os.path.exists(self.file_path)
+
+    @property
+    def file_path(self):
+        return os.path.join(
+            self.store_fs.project.local_fs_path,
+            self.path.strip("/"))
+
+    @property
+    def fs_changed(self):
+        latest_hash = self.latest_hash
+        return (
+            latest_hash is not None
+            and (
+                latest_hash
+                != self.store_fs.last_sync_hash))
+
+    @property
+    def latest_hash(self):
+        if self.exists:
+            return str(os.stat(self.file_path).st_mtime)
+
+    @property
+    def pootle_changed(self):
+        return (
+            self.store is not None
+            and (
+                self.store.get_max_unit_revision()
+                != self.store_fs.last_sync_revision))
+
+    @cached_property
+    def store(self):
+        return self.store_fs.store
+
+    def add(self):
+        logger.debug("Adding file: %s" % self.path)
+        self.store_fs.resolve_conflict = POOTLE_WINS
+        self.store_fs.save()
+
+    def create_store(self):
+        """
+        Creates a ```Store``` and if necessary the ```TranslationProject```
+        parent ```Directories```
+        """
+        store = Store.objects.create_by_path(
+            self.pootle_path,
+            project=self.store_fs.project)
+        self.store_fs.store = store
+        self.store_fs.save()
+        self.__dict__["store"] = self.store_fs.store
+
+    def delete(self):
+        """
+        Delete the file from FS and Pootle
+
+        This does not commit/push
+        """
+        store = self.store
+        if store and store.pk:
+            store.makeobsolete()
+            del self.__dict__["store"]
+        if self.store_fs.pk:
+            self.store_fs.delete()
+        self.remove_file()
+
+    def fetch(self):
+        """
+        Called when FS file is fetched
+        """
+        logger.debug("Fetching file: %s" % self.path)
+        self.store_fs.resolve_conflict = FILE_WINS
+        self.store_fs.save()
+
+    def on_sync(self):
+        """
+        Called after FS and Pootle have been synced
+        """
+        self.store_fs.resolve_conflict = None
+        self.store_fs.staged_for_merge = False
+        self.store_fs.last_sync_hash = self.latest_hash
+        self.store_fs.last_sync_revision = self.store.get_max_unit_revision()
+        self.store_fs.save()
+        logger.debug("File synced: %s" % self.path)
+
+    def pull(self, user=None, merge=False):
+        """
+        Pull FS file into Pootle
+        """
+        current_hash = self.latest_hash
+        last_hash = self.store_fs.last_sync_hash
+        if self.store and last_hash == current_hash:
```

Comment:
I wonder if `self.store is not None` should be used instead. Even more, perhaps it is better to add a `self.store_exists` property (as opposed to `self.exists`) in order to make the code more meaningful.